cmd/tree-main.go

```diff
@@ -0,0 +1,285 @@
+/*
+ * MinIO Client (C) 2019 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/fatih/color"
+	"github.com/minio/cli"
+	"github.com/minio/mc/pkg/console"
+	"github.com/minio/mc/pkg/probe"
+)
+
+const (
+	treeEntry     = "├─ "
+	treeLastEntry = "└─ "
+	treeNext      = "│"
+	treeLevel     = "  "
+)
+
+// Structured message depending on the type of console.
+type treeMessage struct {
+	Entry        string
+	IsDir        bool
+	BranchString string
+}
+
+// Colorized message for console printing.
+func (t treeMessage) String() string {
+	entryType := "File"
+	if t.IsDir {
+		entryType = "Dir"
+	}
+	return fmt.Sprintf("%s%s", t.BranchString, console.Colorize(entryType, t.Entry))
+}
+
+// JSON'ified message for scripting.
+// Does No-op. JSON requests are redirected to `ls -r --json`
+func (t treeMessage) JSON() string {
+	return ""
```

Comment:
You would need to panic() here instead of returning empty string.

Reply:
We should use fatalIf not panic()